crates/ty_python_semantic/src/types.rs

```diff
@@ -6807,6 +6818,26 @@ impl TypeRelation {
     }
 }
 
+/// Represents the result of a subtyping applicability check.
+///
+/// Subtyping only applicable to a pair of types if they are both fully static.
+///
+/// Thus, the wrapped value here will be:
+/// - `None` if the subtyping applicability check has not yet been done.
+/// - `Some(true)` if the subtyping applicability check has been done and both types are fully static.
+/// - `Some(false)` if the subtyping applicability check has been done
+///   and at least one of the types is not fully static.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Default)]
+pub(crate) struct ApplicabilityCheck(Option<bool>);
```

Comment:
We could use [`OnceCell`](https://doc.rust-lang.org/stable/std/cell/struct.OnceCell.html) here to avoid making the `relation` parameter as `mut` across the methods. This would also make `None` unrepresentable.

Reply:
Ugh, the `many_tuple_assignments` benchmark is now actually showing a [1% regression](https://codspeed.io/astral-sh/ruff/branches/alex%2Foptimize-subtyping) after making this change :-(